storage/replica.go

```diff
@@ -701,56 +853,67 @@ func (r *Replica) addWriteCmd(ctx context.Context, args proto.Request, wg *sync.
 
-		// Always push the timestamp forward if there's been a read which
-		// occurred after our txn timestamp.
-		if !rTS.Less(header.Timestamp) {
-			header.Timestamp = rTS.Next()
-		}
-		// If there's a newer write timestamp...
-		if !wTS.Less(header.Timestamp) {
-			// If we're in a txn, we still go ahead and try the write since
-			// we want to avoid restarting the transaction in the event that
-			// there isn't an intent or the intent can be pushed by us.
-			//
-			// If we're not in a txn, it's trivial to just advance our timestamp.
-			if header.Txn == nil {
-				header.Timestamp = wTS.Next()
+			// Always push the timestamp forward if there's been a read which
+			// occurred after our txn timestamp.
+			if !rTS.Less(bArgs.Timestamp) {
+				bArgs.Timestamp = rTS.Next()
+			}
+			// If there's a newer write timestamp...
+			if !wTS.Less(bArgs.Timestamp) {
+				// If we're in a txn, we still go ahead and try the write since
+				// we want to avoid restarting the transaction in the event that
+				// there isn't an intent or the intent can be pushed by us.
+				//
+				// If we're not in a txn, it's trivial to just advance our timestamp.
+				if header.Txn == nil {
+					bArgs.Timestamp = wTS.Next()
+				}
 			}
 		}
 	}
 
+	// Make sure the batch timestamp is copied consistently to each request.
+	// TODO(tschottdorf): self-overlap is real and I think we should deal with
+	// it. Hence, Forward() below instead of overwriting. See setBatchTimestamps.
+	for i := range bArgs.Requests {
+		args := bArgs.Requests[i].GetValue().(proto.Request)
+		args.Header().Timestamp.Forward(bArgs.Timestamp)
+	}
+	r.Unlock()
+
 	defer trace.Epoch("raft")()
 
-	errChan, pendingCmd := r.proposeRaftCommand(ctx, args)
+	errChan, pendingCmd := r.proposeRaftCommand(ctx, bArgs)
 
 	signal()
 
 	// First wait for raft to commit or abort the command.
-	var err error
-	var reply proto.Response
+	var bReply *proto.BatchResponse
 	if err = <-errChan; err == nil {
 		// Next if the command was committed, wait for the range to apply it.
 		respWithErr := <-pendingCmd.done
-		reply, err = respWithErr.Reply, respWithErr.Err
+		bReply, err = respWithErr.Reply, respWithErr.Err
 	}
 
-	// As for reads, update timestamp cache with the timestamp
-	// of this write on success. This ensures a strictly higher
-	// timestamp for successive writes to the same key or key range.
-	r.endCmd(cmdKey, args, err, false /* !readOnly */)
-	return reply, err
+	r.endCmds(cmdKeys, bArgs, err)
+	return bReply, err
 }
 
 // proposeRaftCommand prepares necessary pending command struct and
 // initializes a client command ID if one hasn't been. It then
 // proposes the command to Raft and returns the error channel and
 // pending command struct for receiving.
-func (r *Replica) proposeRaftCommand(ctx context.Context, args proto.Request) (<-chan error, *pendingCmd) {
+func (r *Replica) proposeRaftCommand(ctx context.Context, bArgs *proto.BatchRequest) (<-chan error, *pendingCmd) {
```

Comment:
gave this a shot and various things panic, I think it's because we advance the request header timestamp in some places and those updates must change the "original". I added a TODO in `AddCmd` to revisit this when the batch stuff is in - it'll be good to have a little more control over when the request is mutated. With batches, we'll eventually get rid of all mutable data on the individual requests (at least in `storage`) so not passing pointers needlessly should give us some safety there eventually.